### PR TITLE
External Authentication Feature

### PR DESCRIPTION
**Pull request by MikeAlhayek**
### External Authentication Feature

We've introduced a new feature called **External Authentication**. This feature has been separated from the existing **Users** feature to improve dependency management and to provide an option to disable external authentication by default. As a result, the **User Registration** feature no longer needs to be enabled unless you specifically want to allow site registration.

This feature is only available on-demand and cannot be manually enabled or disabled. It is automatically enabled when a feature requiring external authentication is activated.

We included a fast-forward migration to enable this feature automatically if previously needed. The following settings have been relocated to new classes as part of this update:

- The properties `UseScriptToSyncRoles` and `SyncRolesScript` have been moved from `LoginSettings` to the new `ExternalLoginSettings` class.
- The property `UseExternalProviderIfOnlyOneDefined` has also been moved from `LoginSettings` to `ExternalLoginSettings`.

In addition, several properties have been moved from `RegistrationSettings` to a new `ExternalRegistrationSettings` class:

- `NoPasswordForExternalUsers` is now `ExternalRegistrationSettings.NoPassword`.
- `NoUsernameForExternalUsers` is now `ExternalRegistrationSettings.NoUsername`.
- `NoEmailForExternalUsers` is now `ExternalRegistrationSettings.NoEmail`.
- `UseScriptToGenerateUsername` is now `ExternalRegistrationSettings.UseScriptToGenerateUsername`.
- `GenerateUsernameScript` is now `ExternalRegistrationSettings.GenerateUsernameScript`.

Also, note the following updates in `ExternalLoginSettings`:

- `UseScriptToSyncRoles` has been renamed to `ExternalLoginSettings.UseScriptToSyncProperties`.
- `SyncRolesScript` has been renamed to `ExternalLoginSettings.SyncPropertiesScript`.

!!! note
    When updating recipes to configure `LoginSettings` or `RegistrationSettings`, ensure that the settings reflect the new class structure.

### User Registration Feature

The **User Registration** feature is no longer required if you only want to enable external authentication. 

The following properties of `RegistrationSettings` are now deprecated and will be removed in the next major release:

- `UsersCanRegister`
- `NoPasswordForExternalUsers`
- `NoUsernameForExternalUsers`
- `NoEmailForExternalUsers`
- `UseScriptToGenerateUsername`
- `GenerateUsernameScript`

Previously, the `UsersCanRegister` property controlled which types of registration were allowed. With this update, this property is obsolete and will be removed in a future release. To enable site registration now, simply activate the **User Registration** feature.

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountBaseController.cs [hunk 1]
```diff
@@ -36,4 +37,25 @@ protected IActionResult RedirectToLocal(string returnUrl)
...
     }
+
+    protected void CopyTempDataErrorsToModelState()
+    {
+        foreach (var errorMessage in TempData.Where(x => x.Key.StartsWith("error")).Select(x => x.Value.ToString()))
+        {
+            ModelState.AddModelError(string.Empty, errorMessage);
+        }
+    }
+
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountBaseController.cs:48]:
```suggestion
```

File: src/OrchardCore.Modules/OrchardCore.Users/Extensions/ControllerExtensions.cs [hunk 2]
```diff
@@ -60,7 +60,7 @@ internal static async Task<IUser> RegisterUser(this Controller controller, Regis
 
         var settings = await controller.ControllerContext.HttpContext.RequestServices.GetRequiredService<ISiteService>().GetSettingsAsync<RegistrationSettings>();
 
-        if (settings.UsersCanRegister != UserRegistrationType.NoRegistration)
+        if (settings.AllowSiteRegistration)
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Users/Extensions/ControllerExtensions.cs:63]: Have to add "or allow external login" or probably better trigger an handler that would check if registration should be enabled or not

**Review by MikeAlhayek — COMMENTED** [review]
Comments to myself.

**Comment by MikeAlhayek** [comment]
@mvarblow if you are able to, please test out this PR.

File: src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalAuthenticationSettings.Edit.cshtml
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalAuthenticationSettings.Edit.cshtml:1]: This and src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalUserLoginSettings.Edit.cshtml would look better under a new title, just as the registration settings ("External Authentication Settings when registering with external authentication providers").

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 3]
```diff
@@ -301,668 +253,23 @@ public async Task<IActionResult> ChangePassword(ChangePasswordViewModel model, s
     public IActionResult ChangePasswordConfirmation()
         => View();
 
-    [HttpPost]
-    [AllowAnonymous]
-    [ValidateAntiForgeryToken]
-    public IActionResult ExternalLogin(string provider, string returnUrl = null)
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:307]: So the methods were all moved from the AccountController. So these methods are not part of the new feature.

File: src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml [hunk 4]
```diff
@@ -9,24 +9,6 @@
...
-    <span asp-validation-for="UsersCanRegister"></span>
-
-    <div class="hint">
-        <ul>
-            <li>@T["when 'No Registration' is selected, no new users can register"]</li>
-            <li>@T["when 'Allow Registration' is selected, a link is displayed to allow the users to register on the site. If there are external providers, once authenticated the user can register"]</li>
-            <li>@T["when 'Allow Only External Users' is selected, only users authenticated with external providers can register on the site"]</li>
-        </ul>
-    </div>
-</div>
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml:28]: look at the release notes to see the change. `UserRegistrationType` is no longer available. Instead you can set the new `AllowSiteRegistrations` to true instead. if you need `AllowOnlyExternalUsers` then just enable the new feature instead. so we don't have the types anymore.

File: src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalAuthenticationSettings.Edit.cshtml
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalAuthenticationSettings.Edit.cshtml:1]: I updated both settings to have this as a header

```
<h5 class="mb-3">@T["Configuring External Authentication for User Registration with Third-Party Providers"]</h5>
```

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:307]: There's a lot of code in this and other classes, and templates and else, that are removed, and have been factored out to the new feature and various base/helper classes. My question is: did you change anything substantial in the new code files, or did you only basically copied over the code? Because any changes are not visible, unless I visually match every piece of code (or copy them over and see if anything changed) one by one, what I'd avoid.

File: src/docs/releases/2.1.0.md [hunk 5]
```diff
@@ -4,6 +4,31 @@ Release date: Not yet released
...
+- `UseScriptToGenerateUsername` is now `ExternalAuthenticationSettings.UseScriptToGenerateUsername`.
+- `GenerateUsernameScript` is now `ExternalAuthenticationSettings.GenerateUsernameScript`.
+
+The `UsersCanRegister` property in the `RegistrationSettings` class has been marked obsolete, and a new property, `AllowSiteRegistration`, has been introduced. To enable site registration, set `AllowSiteRegistration` to `true`. For external registration, ensure the **External User Authentication** feature is enabled.
+
+!!! note
+    When configuring `LoginSettings` or `RegistrationSettings` in a recipe, ensure that the recipe's properties are updated to reflect the new settings classes.
+
+!!! note
+    When the **External User Authentication** feature is enabled, the `/Register` endpoint will always become available.
```
Comment (Piedone) [inline src/docs/releases/2.1.0.md:30]: Move this up after the first two paragraphs, because this is more important than the property changes that should affect very few people (only when you programmatically interact with these settings).

File: src/docs/releases/2.1.0.md [hunk 6]
```diff
@@ -4,6 +4,31 @@ Release date: Not yet released
 
 ## Change Logs
 
+### External User Authentication Feature
+
+A new feature, **External User Authentication**, has been added. This feature was separated from the existing **Users** feature to provide better dependency management and offer an option to disable external authentication by default.
+
+After upgrading the site, a fast-forward option allows you to enable this feature automatically when needed. As a result, the following settings have been relocated to new classes:
```
Comment (Piedone) [inline src/docs/releases/2.1.0.md:11]:
```suggestion
After upgrading the site, a fast-forward option allows you to enable this feature automatically when needed. No action is needed, everything will be migrated automatically.

As a result, the following settings have been relocated to new classes:
```

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:307]: No. The code was mainly copied from one place to another "almost as-is". The main difference is the introduction of the options classes. We no longer access the settings from the SiteService. Instead we use in memory Options as these options would change depending on the enabled feature. For example, the RegaitrationOptions is configured when the registration feature is enabled. Also we do similar thing when external feature is enabled, we always ensure that site registration is enabled to allow external users to login.

[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:307]: Also, we no longer need to check if the registration feature is enabled as we did previously. With the new approach this is better because we are using feature dependency instead of manually checking for enabled feature which is what we did in the past

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml:28]: Yeah, I see the options classes, but just wanted to make sure if e.g. `if`s where switched around or something. Great, then!

[the remark this answers is not in the document]
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml:28]: According to [ln 48 in `ExternalUserMigrations`](https://github.com/OrchardCMS/OrchardCore/pull/16721/files#diff-02b9a309678bcdd3a6949d887f2a512e491f83d871c19b8557b4cab196bd4bb3R48) just enabling the new feature should be enough for something like Entra ID to work, but it isn't. I don't see the equivalent configuration of what was removed above.

I.e.:

- What should I configure _from the UI_ to allow/disallow user registration (while keeping the Users Registration feature enabled, as before)?
- What should I configure _from the UI_ to allow/disallow Entra ID login (after enabling the Entra ID feature)?

I see neither being possible, only from code.
 
Also, previously, only enabling the Users Registration feature wasn't enough to allow users to register, because you had to set this option. While I don't particularly agree with this UX (it being superfluous with the feature enable), now just enabling the Entra ID feature is also enough to register a new local user (i.e. just enabling Entra ID enables local, not just external, user registration too).

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:307]: Yeah, I see the options classes, but just wanted to make sure if e.g. `if`s where switched around or something. Great, then!

File: src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml (hunk 4 above)
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml:28]: @Piedone 

> What should I configure from the UI to allow/disallow user registration (while keeping the Users Registration feature enabled, as before)?

There is a new AllowSiteRegistration on the registration settings on the registration UI that you can set.

> What should I configure from the UI to allow/disallow Entra ID login (after enabling the Entra ID feature)?

You should not need anything. Enabling entraID will automatically enable the new feature which will provide everything needed for the external authentication to work "including" the /Register route "without needing to enable the registration feature".

File: src/docs/releases/2.1.0.md (hunk 5 above)
Comment (MikeAlhayek) [reply src/docs/releases/2.1.0.md]: done

File: src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml (hunk 4 above)
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml]: I was missing the `AllowSiteRegistration` on the UI so I added it.

[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml]: @Piedone I think for Entra ID to work, you'll need to call [ConfigureMicrosoftAccountSettings](https://github.com/OrchardCMS/OrchardCore/blob/b3c4b4689135498b0afcf687b801b9d149e6f5b3/src/OrchardCore.Modules/OrchardCore.Microsoft.Authentication/Extensions/OrchardCoreBuilderExtensions.cs#L9-L19) extensions some where in your code. Not great design but this is an existing behavior "not related to this PR". Can you try calling `ConfigureMicrosoftAccountSettings()` on OC builder and see if you'll get the button?

[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml]: It's also mentioned in the docs.

https://docs.orchardcore.net/en/latest/reference/modules/Microsoft.Authentication/#microsoft-account-microsoft-entra-id-azure-active-directory-settings-configuration

[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml]: Have some minor changes to do in the morning

**Comment by mvarblow** [comment]
@MikeAlhayek, I pulled down this branch to evaluate whether it fixes the same problem I was attempting to address in #16716. First, on the main branch, I configured Entra ID authentication and user registration and confirmed that they were working. Then I checked out this PR branch. The first thing I noticed was that the validation on the login page seemed broken. When I click the "Work Account" button (the name I gave to the Entra ID provider), it complains about the user name and password fields being blank.

image

In fact, even if I fill in those fields before clicking the "Work Account" button to sign in with Entra ID, I still get those same validation errors. I can no longer sign in using Entra ID. Local sign-in (if I press the "Log in" button) does still work.

**Comment by mvarblow** [comment]
@MikeAlhayek, I see you added a couple of commits this morning that I didn't have. I've pulled the latest and now am able to sign in using Entra ID. I'll continue testing...

**Comment by mvarblow** [comment]
@MikeAlhayek This does not seem to address the same problem that I fixed in https://github.com/OrchardCMS/OrchardCore/pull/16716. First, on the main branch, I configured Entra ID authentication and user registration and confirmed that they were working. Then I disabled User Registration. I reproduced the bug (external user registration is still allowed!). Then I switched to this PR branch. The External User Authentication feature was automatically enabled, as it should be because I had Microsoft Entra ID Authentication enabled (to allow external **sign-in**). When I tried to sign in using a new Entra ID account, even though the User Registration feature is disabled (see above), I was still allowed to register via the ExternalLoginCallback endpoint.

**Comment by MikeAlhayek** [comment]
@mvarblow please pull the latest and try.

With the latest change, there is no registration type settings. With the current PR, if you enabled **User Registration** feature, then users can register users using the form "not external users". If you want to enable external authentication, you'll need to enable at least one of the authentication provider like Entra ID feature.

So `RegistrationSettings` is now independent of External site registration. If you enable external authentication, you'll always expose a way to register external users (otherwise external authentication will not work for new users).

File: src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml (hunk 4 above)
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml:28]: fixed.

**Comment by mvarblow** [comment]
@MikeAlhayek I saw that you pushed a couple of additional commits for this PR. I just pulled the latest (97b910b42dc940a5e256269cb0488d8f9ea321e5). Now if I run my test I get this NRE on startup after switching to the PR branch:

image

image

As before, my test plan looks like this:

1. Start the site (in VS) and complete setup using the Agency recipe.
2. Enable Microsoft Entra ID authentication and User Registration (external users only).
3. Disable User Registration.
4. Confirm bug (external registration still allowed).
5. Switch to the PR branch (ma/external-login-feature) and start the site.
6. Confirm bug is fixed, or not (registration should not be allowed).

The exception above was thrown on step 5. This is new behavior with the latest commits.

**Comment by MikeAlhayek** [comment]
@mvarblow thank you for testing. The NRE should now be fixed.

**Comment by mvarblow** [comment]
@MikeAlhayek The NRE is fixed. However, I'm still able to register even though the User Registration feature is disabled. 

image

**Comment by MikeAlhayek** [comment]
@mvarblow 
Yes this is expected. This type of registration is only for external users and can't be disabled since it's a critical aspect of the external login. However, you should not be able to click on "Register" on the site and register a new user using a form. In other words, the /register route will not be available unless you enabled the registration feature.

**Comment by mvarblow** [comment]
@MikeAlhayek So this is an expected change in behavior? If so, I would suggest that this change should not got into 2.1, but wait for 3.0. This would be a breaking change for us. We have a few different configurations that our clients use -- one very common configuration is that external sign-in is enabled but user registration is disabled. An administrator must provision all users, including external users. This is supported in Orchard 2.0 (and earlier). In those releases, if I disable the User Registration feature then it also disables registration for external users. 

(The bug in earlier releases that I was trying to fix is very subtle. If I left user registration disabled when I set up the site then user registration was blocked for both local and external users. Again, we counted on this behavior. However, if enabled user registration and set the UsersCanRegister property to anything other than NoRegistration _before_ disabling the user registration feature then regular user registration was correctly disabled but external user authentication was not.)

If we're going to change the behavior as you suggest, could we at least control it via a setting in the external authentication feature, give admins the option to turn off registration for external users?

**Comment by MikeAlhayek** [comment]
> So this is an expected change in behavior? If so, I would suggest that this change should not got into 2.1, but wait for 3.0. This would be a breaking change for us. We have a few different configurations that our clients use -- one very common configuration is that external sign-in is enabled but user registration is disabled. An administrator must provision all users, including external users. This is supported in Orchard 2.0 (and earlier). In those releases, if I disable the User Registration feature then it also disables registration for external users.

In your case, how can new user login for the very first time if they can't login? Typically with external authentication, new users will need to be registered into the app so a local account is created.

**Comment by mvarblow** [comment]
@MikeAlhayek We have two supported options - 

(1) The admin pre-provisions a user with the correct email address; when the user signs in for the first time OrchardCore finds the user by email; the user is prompted to enter their password (which the admin provided them after provisioning the account) to link the external ID. This is standard Orchard Core. 

(2) We added our own module that gives the admin the ability to manage external IDs for users via the OrchardCore dashboard. The admin pre-provisions the user and also adds the external ID (supplying the Entra ID user ID). This is what most of our clients do because it doesn't require the user to enter an Orchard password on first sign-in. They usually have local sign-in disabled, so users don't normally need to know (or have) a local password.

**Comment by MikeAlhayek** [comment]
@mvarblow I added a new setting `DisableNewRegistrations` to the `ExternalRegistrationSettings` which will allow you to achieve your use case.

In your case, the migration should set `DisableNewRegistrations` to true, it will disable "User Registrations" feature and the "External Authentication" will be auto enabled by Microsoft EntraID feature.

**Comment by MikeAlhayek** [comment]
@mvarblow just a suggestion to your approach (which seems to require lots of work from admin). I suggest you test the following setup instead:

1. Don't provision the users manually.
2. Enable external registration
3. On the `ExternalRegistrationSettings` use the following properties:
   - Use the `SyncPropertiesScript` to sync username/id/email with EntraID.
   - Set `UseGenerateUsernameScript` to true
   - Use the `GenerateUsernameScript` to generate a user name based on the claims from Entra ID.
   - Set `NotPassword` to true.
   - Set `NoUsername` to true.
   - Set `NoEmail` to true.
 4. Then optionally, on the `ExternalLoginSettings` you can use these properties if needed
    - Set `UseScriptToSyncProperties` to true.
    - Set `SyncPropertiesScript` to set other properties if needed like roles, claims, phone number or anything else that you may need.

**Comment by mvarblow** [comment]
@MikeAlhayek Yes, we've considered that approach. Unfortunately, there are settings that need to be defined for users before they use the app which can't easily be pulled from Entra ID. Entra ID could be extended to contain the necessary information but that's often not practical for our clients. I'm sure we will revisit this in the future though.

Regarding your latest commit, I appreciate you adding that setting. That's perfect for us and gives the flexibility to keep things working as before. Thanks!

The migration didn't seem to set the new property as I expected. I did have to, after swapping to the PR branch, manually check that option. I think you expected it to be set by the migration?

image

And I noticed a small cosmetic issue. The notification (when I sign in using a new external user after I checked the Disable New External Authentication box) spans the whole page while the rest of the form is width limited. This looks a bit sloppy. But it does not appear related to this merge request.

image

**Comment by mvarblow** [comment]
@MikeAlhayek I see what's happening in the migration. It looks fine. It's not detecting my special case (you enable user registration and allow external user registration; then later disable user registration). But it at least gives me an easy way to explicitly turn on and off external user registration. This is great! Thanks.

**Comment by MikeAlhayek** [comment]
@mvarblow yes I think it's because you already run the same migration so it wont run again. You can modify the migration document if you need manually in the database to force it to run again.

If you see no issue with this PR (since you are testing it, please approve it) just so we have you as a tester

File: src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml (hunk 4 above)
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml:28]: @Piedone `AllowSiteRegistration` was removed. If you need to enable user registration, simply "User Registration" feature. If you enable any identity-provider feature, the External Authentication feature will be enabled by dependency. By default, we enable registration for external users. However, I added a new setting that allows us to disable external registration "if needed".

File: src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalAuthenticationSettings.Edit.cshtml
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalAuthenticationSettings.Edit.cshtml:1]: I also group settings using tabs. So that external authentication settings are in their own tab.

File: src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml:28]: You're right about `ConfigureAzureADSettings` (`ConfigureMicrosoftAccountSettings` is for personal Microsoft accounts), I missed that. With that, Entra ID config and login works.

However, the previous behavior of not allowing registration by default if Users Registration is enabled, only by flipping a switch, should be retained, otherwise this would be a breaking change. We can remove it in 3.0.

[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml:28]: @Piedone 
This is not a breaking change. The migration already provides settings updates to ensure we don't have any breaking changes.

In the migration, if you already prevent registration, then we disable the User registration feature and enable "DisableExternalRegustration" settings. The External authentications feature will be enabled if any of the providers are enabled. 

If you are running a new tenant, the. DisableExternalRegisteration will be set to false by default.

File: src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalLoginSettings.Edit.cshtml [hunk 7]
```diff
@@ -52,34 +65,34 @@
...
-                    valueType: string;
-                    type: string;
-                    value: string;
-                    subject: string;
-                    issuer: string;
-                    originalIssuer: string;
-                    properties: { [key: string]: string };
-                }]
-            }`
+                                                            * Use the loginProvider and externalClaims properties of the context variable to inspect
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalLoginSettings.Edit.cshtml:68]: I guess the indentation is not deliberate:

{1473DFF0-C1B2-4EF4-8EB7-A9C91A10A06C}

File: src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalRegistrationSettings.Edit.cshtml [hunk 8]
```diff
@@ -11,30 +11,42 @@
 
 <h5 class="mb-3">@T["Configuring External Authentication for User Registration with Third-Party Providers"]</h5>
 
+<div class="mb-3" asp-validation-class-for="DisableNewRegistrations">
+    <div class="form-check">
+        <input type="checkbox" class="form-check-input" asp-for="DisableNewRegistrations" />
+        <span asp-validation-for="DisableNewRegistrations"></span>
+        <label class="form-check-label" asp-for="DisableNewRegistrations">@T["Disable new external registration"]</label>
+        <span class="hint dashed">@T["If selected, new users will not be able to register using external authentication. This means users without a local account will be unable to authenticate."]</span>
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalRegistrationSettings.Edit.cshtml:19]:
```suggestion
        <span class="hint dashed">@T["If selected, only users with existing local accounts will be able to authenticate, and new users will not be able to register using external authentication. Users without a local account will not be able to authenticate."]</span>
```

File: src/OrchardCore/OrchardCore.Abstractions/Shell/Extensions/ShellFeaturesManagerExtensions.cs [hunk 9]
```diff
@@ -34,6 +34,35 @@ public static async Task EnableFeaturesAsync(this IShellFeaturesManager shellFea
         await shellFeaturesManager.EnableFeaturesAsync(featuresToEnable, force: false);
     }
 
+    public static async Task UpdateFeaturesAsync(this IShellFeaturesManager shellFeaturesManager, IEnumerable<string> featureIdsToDisable, IEnumerable<string> featureIdsToEnable)
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Abstractions/Shell/Extensions/ShellFeaturesManagerExtensions.cs:37]: Why is this necessary? It's not used.

File: src/docs/releases/2.1.0.md [hunk 10]
```diff
@@ -4,30 +4,49 @@ Release date: Not yet released
...
+We've introduced a new feature called **External Authentication**. This feature has been separated from the existing **Users** feature to improve dependency management and to provide an option to disable external authentication by default. As a result, the **User Registration** feature no longer needs to be enabled unless you specifically want to allow site registration.
 
-After upgrading the site, a fast-forward option allows you to enable this feature automatically when needed. As a result, the following settings have been relocated to new classes:
+This feature is only available on-demand and cannot be manually enabled or disabled. It is automatically enabled when a feature requiring external authentication is activated.
 
-- The properties `UseScriptToSyncRoles` and `SyncRolesScript` have been moved from `LoginSettings` to a new class, `ExternalUserRoleLoginSettings`.
-- The property `UseExternalProviderIfOnlyOneDefined` has been moved from `LoginSettings` to `ExternalUserLoginSettings`.
+We included a fast-forward migration to enable this feature automatically if previously needed. The following settings have been relocated to new classes as part of this update:
```
Comment (Piedone) [inline src/docs/releases/2.1.0.md:13]:
```suggestion
We included a fast-forward migration to enable this feature automatically if previously needed. No action is needed from site administrators.

The following settings have been relocated to new classes as part of this update:
```

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Users/DataMigrations/ExternalAuthenticationMigrations.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            if (enumValue is not null && enumValue != 1)
+            {
+                var featuresManager = scope.ServiceProvider.GetRequiredService<IShellFeaturesManager>();
+
+                if (await featuresManager.IsFeatureEnabledAsync(UserConstants.Features.UserRegistration))
+                {
+                    return;
+                }
+
+                await featuresManager.DisableFeaturesAsync(UserConstants.Features.UserRegistration);
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/DataMigrations/ExternalAuthenticationMigrations.cs:56]: Yup. It should be negated. I'll update

File: src/OrchardCore/OrchardCore.Abstractions/Shell/Extensions/ShellFeaturesManagerExtensions.cs (hunk 9 above)
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Abstractions/Shell/Extensions/ShellFeaturesManagerExtensions.cs:37]: I needed it at some point during this PR. It's probably not a bad idea to keep it as a helpful extension

File: src/OrchardCore/OrchardCore.Users.Core/Models/ExternalLoginOptions.cs [hunk 12]
```diff
@@ -1,6 +1,6 @@
 namespace OrchardCore.Users.Models;
 
-public class ExternalUserLoginOptions
+public class ExternalLoginOptions
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Users.Core/Models/ExternalLoginOptions.cs:3]: no. It should use `ExternalRegistrationSettings`

File: src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalLoginSettings.Edit.cshtml (hunk 7 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Views/ExternalLoginSettings.Edit.cshtml]: yeah VS did some magic here. Fixed it.

File: src/OrchardCore.Modules/OrchardCore.Users/DataMigrations/ExternalAuthenticationMigrations.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/DataMigrations/ExternalAuthenticationMigrations.cs]: updated

**Review by Piedone — APPROVED** [review]
@mvarblow could you please do a final check?

**Comment by mvarblow** [comment]
@MikeAlhayek Actually, I'm reverting the site each time I test the PR. So the migration is running. I have confirmed that in the debugger. Remember, in my case the UserRegistration feature is already disabled. The problem I experienced in Orchard 2.0 is that external registration is still allowed after I disabled the UserRegistration feature. When I use your PR, the migration runs and sees that the UsersCanRegister property value is 2 (how it was set before I disabled the UserRegistration feature previously). But it also sees that the UserRegistration is disabled, so it doesn't need to disable. To automatically fix the problem I pointed out originally, the migration would have to see that the UserRegistration feature is disabled and, regardless of the old UserCanRegister property value, decide to set DisableNewRegistrations to true.

As it is, your PR helps me. But I have to manually make that settings change after deploying the updated OrchardCore version.

To automate the fix-up I think you'd need to change:

                DisableNewRegistrations = enumValue == 0,

to

                DisableNewRegistrations = enumValue == 0 || !await featuresManager.IsFeatureEnabledAsync(UserConstants.Features.UserRegistration),

**Comment by MikeAlhayek** [comment]
@mvarblow I agree with the request and made the change. Can you please confirm that it is working for you?

**Comment by mvarblow** [comment]
@MikeAlhayek Looks good! Thanks for this!
